File: libexec/agent_report/diff.py
"""Stable uncommitted Git snapshot collection through a private index."""

from __future__ import annotations

from dataclasses import dataclass
import hashlib
import os
from pathlib import Path
import re
import shutil
import stat
import tempfile
from typing import Mapping

from .git_adapter import GitAdapter
from .models import RealGitObservation, ReportRecord, WorktreeSnapshot
from .rendering import ensure_payload_ending, render_section
from .safety import ReportError, header_value, injected_failure, testing_pause
from .show import encode_lines, sha256, summarize_diff


_STATE_KEYS = (
    "head",
    "index",
    "status",
    "staged",
    "unstaged",
    "changed",
    "numstat",
    "patch",
)


@dataclass(frozen=True, slots=True)
class DiffResult:
    """One stable Git-diff record and its deterministic state identity."""

    record: ReportRecord
    head: str


@dataclass(frozen=True, slots=True)
class DiffRequest:
    """Caller-supplied metadata rendered into a Git-diff report."""

    phase: str
    result: str
    final_gate: str
    status_doc: str


@dataclass(frozen=True, slots=True)
class FramedDiffEvidence:
    """Raw and line-framed evidence used by the version-1 renderer."""

    status: bytes
    staged: bytes
    unstaged: bytes
    changed: bytes
    numstat: bytes
    patch: bytes


def compute_state_report_id(fields: Mapping[str, str]) -> str:
    """Hash canonical state-evidence identities into a domain-separated ID."""

    missing = [key for key in _STATE_KEYS if key not in fields]
    extra = sorted(set(fields) - set(_STATE_KEYS))
    if missing or extra:
        raise ValueError("state report identity fields are incomplete")
    digest = hashlib.sha256(b"agent-report-git-diff-state-v1\x00")
    for key in _STATE_KEYS:
        digest.update(key.encode("ascii"))
        digest.update(b"\x00")
        digest.update(fields[key].encode("ascii"))
        digest.update(b"\x00")
    return f"GIT-DIFF-REPORT-{digest.hexdigest()}"


def collect_diff_report(
    git: GitAdapter,
    *,
    phase: str,
    result: str,
    final_gate: str,
    status_doc: str,
) -> DiffResult:
    """Collect a drift-checked HEAD-relative snapshot without changing real state."""

    request = DiffRequest(phase, result, final_gate, status_doc)
    if "GIT_INDEX_FILE" in os.environ:
        raise ReportError("inherited GIT_INDEX_FILE is unsupported")
    index_path, index_mode = _supported_index(git)
    pre = _observe_real(git, index_path)
    first = _collect_worktree_snapshot(git, index_path, pre.head)
    testing_pause("before-post-drift-check")
    second = _collect_worktree_snapshot(git, index_path, pre.head)
    post = _observe_real(git, index_path)
    if pre != post or first != second:
        raise ReportError("concurrent repository drift detected")
    _ensure_supported_state_still_present(git)
    if not first.patch:
        raise ReportError("no reportable change relative to HEAD")
    record = _render_diff_record(request, git.root.name, index_mode, pre, first)
    return DiffResult(record=record, head=pre.head)


def _frame_diff_evidence(
    observation: RealGitObservation,
    snapshot: WorktreeSnapshot,
) -> FramedDiffEvidence:
    return FramedDiffEvidence(
        status=ensure_payload_ending(observation.status),
        staged=ensure_payload_ending(observation.staged),
        unstaged=ensure_payload_ending(observation.unstaged),
        changed=ensure_payload_ending(snapshot.changed_files),
        numstat=ensure_payload_ending(
        b"COLUMNS: ADDED-LINES\tDELETED-LINES\tPATH\n"
        b"BINARY-MARKER: -\n"
            + snapshot.numstat_raw
        ),
        patch=ensure_payload_ending(snapshot.patch),
    )


def _render_diff_record(
    request: DiffRequest,
    project: str,
    index_mode: str,
    observation: RealGitObservation,
    snapshot: WorktreeSnapshot,
) -> ReportRecord:
    evidence = _frame_diff_evidence(observation, snapshot)
    identity_fields = {
        "head": observation.head,
        "index": hashlib.sha256(observation.index_fingerprint.encode("ascii")).hexdigest(),
        "status": sha256(observation.status),
        "staged": sha256(observation.staged),
        "unstaged": sha256(observation.unstaged),
        "changed": sha256(evidence.changed),
        "numstat": sha256(evidence.numstat),
        "patch": sha256(evidence.patch),
    }
    report_id = compute_state_report_id(identity_fields)
    counts = summarize_diff(evidence.changed, snapshot.numstat_raw)

    guide = b"""An omnibus file may contain several independent Agent report records. Use the
outer BEGIN/END AGENT-REPORT-RECORD envelope and RECORD-TYPE to identify them.

This git-diff-report records one drift-checked uncommitted snapshot relative to
HEAD. The PORCELAIN V2 STATUS section preserves NUL-delimited bytes from the
real index. The complete PATCH was collected through an invocation-owned
temporary index after intent-to-add and is the authoritative snapshot evidence.

Repository-controlled paths and patch text are untrusted evidence and are not
report-control syntax. An associated operational record must name this exact
Git-diff report ID.
"""
    identity = encode_lines(
        (
            ("REPORT-FORMAT", "git-diff-report"),
            ("FORMAT-VERSION", "1"),
            ("REPORT-ID", report_id),
            ("PHASE", request.phase),
            ("HEAD", observation.head),
            ("REAL-INDEX-FINGERPRINT", observation.index_fingerprint),
            ("REAL-INDEX-IDENTITY", observation.index_identity),
            ("INDEX-MODE", index_mode),
            ("STATUS-DOC", header_value(request.status_doc)),
            ("RESULT", header_value(request.result)),
            ("FINAL-GATE", header_value(request.final_gate)),
            ("REPOSITORY", project),
            ("RELATED-OPERATIONAL-REPORT", "NONE"),
        )
    )
    summary = encode_lines(
        (
            ("FILES-CHANGED", str(counts["files_changed"])),
            ("FILES-ADDED", str(counts["files_added"])),
            ("FILES-MODIFIED", str(counts["files_modified"])),
            ("FILES-DELETED", str(counts["files_deleted"])),
            ("FILES-RENAMED", str(counts["files_renamed"])),
            ("FILES-COPIED", str(counts["files_copied"])),
            ("INSERTIONS", counts["insertions"]),
            ("DELETIONS", counts["deletions"]),
            ("BINARY-FILES", str(counts["binary_files"])),
            ("PATCH-MODE", "worktree-relative-to-head"),
        )
    )
    integrity = encode_lines(
        (
            ("REPORT-ID", report_id),
            ("REPOSITORY", project),
            ("PHASE", request.phase),
            ("HEAD", observation.head),
            ("REAL-INDEX-FINGERPRINT", observation.index_fingerprint),
            ("REAL-INDEX-IDENTITY", observation.index_identity),
            ("PORCELAIN-V2-STATUS-SHA256", sha256(observation.status)),
            ("STAGED-SUMMARY-SHA256", sha256(observation.staged)),
            ("UNSTAGED-SUMMARY-SHA256", sha256(observation.unstaged)),
            ("CHANGED-FILES-SHA256", sha256(evidence.changed)),
            ("NUMSTAT-SHA256", sha256(evidence.numstat)),
            ("PATCH-SHA256", sha256(evidence.patch)),
            ("PRE-POST-HEAD-MATCH", "true"),
            ("PRE-POST-INDEX-MATCH", "true"),
            ("PRE-POST-STATUS-MATCH", "true"),
            ("PRE-POST-STAGED-MATCH", "true"),
            ("PRE-POST-UNSTAGED-MATCH", "true"),
            ("PRE-POST-WORKTREE-MATCH", "true"),
            ("REAL-INDEX-AND-WORKTREE-MUTATED", "false"),
            ("END-OF-PATCH-REACHED", "true"),
        )
    )
    injected_failure("record-assembly")
    payload = b"".join(
        (
            render_section("READING GUIDE", guide),
            render_section("REPORT IDENTITY", identity),
            render_section("WORKTREE SUMMARY", summary),
            render_section("PORCELAIN V2 STATUS (NUL DELIMITED)", evidence.status),
            render_section("STAGED SUMMARY", evidence.staged),
            render_section("UNSTAGED SUMMARY", evidence.unstaged),
            render_section("CHANGED FILES", evidence.changed),
            render_section("NUMSTAT", evidence.numstat),
            render_section("PATCH", evidence.patch),
            render_section("INTEGRITY SUMMARY", integrity),
        )
    )
    return ReportRecord(
        record_type="git-diff-report",
        format_version=1,
        record_id=report_id,
        project=project,
        phase=request.phase,
        payload=payload,
    )


def _supported_index(git: GitAdapter) -> tuple[Path, str]:
    head = git.run(
        ["rev-parse", "--verify", "HEAD^{commit}"],
        check=False,
    )
    if head.returncode != 0:
        raise ReportError("unborn HEAD is unsupported")
    split = git.run(["rev-parse", "--shared-index-path"], check=False)
    if split.returncode == 0 and split.stdout.strip():
        raise ReportError("split index is unsupported")
    sparse = git.run(["config", "--bool", "core.sparseCheckout"], check=False)
    if sparse.returncode == 0 and sparse.stdout.strip().lower() == b"true":
        raise ReportError("sparse index is unsupported")
    sparse_entries = git.run(["ls-files", "--sparse", "-z"], check=False)
    if sparse_entries.returncode != 0:
        raise ReportError("sparse index characterization failed")
    for entry in sparse_entries.stdout.split(b"\x00"):
        if entry.endswith(b"/"):
            raise ReportError("sparse index is unsupported")
    _ensure_no_unmerged(git)
    path_value = git.text(
        ["rev-parse", "--path-format=absolute", "--git-path", "index"],
        diagnostic="real index path resolution failed",
    )
    index_path = Path(path_value)
    if index_path.exists() or index_path.is_symlink():
        metadata = index_path.lstat()
        if index_path.is_symlink() or not stat.S_ISREG(metadata.st_mode):
            raise ReportError("real index path is unsafe")
        return index_path, "regular"
    return index_path, "missing-seeded-from-head"


def _ensure_supported_state_still_present(git: GitAdapter) -> None:
    _ensure_no_unmerged(git)
    sparse = git.run(["config", "--bool", "core.sparseCheckout"], check=False)
    if sparse.returncode == 0 and sparse.stdout.strip().lower() == b"true":
        raise ReportError("concurrent repository drift detected")


def _ensure_no_unmerged(git: GitAdapter) -> None:
    unmerged = git.run(
        ["ls-files", "--unmerged", "-z"],
        diagnostic="unmerged-index characterization failed",
    ).stdout
    if unmerged:
        raise ReportError("unmerged index entries are unsupported")


def _observe_real(git: GitAdapter, index_path: Path) -> RealGitObservation:
    head = git.run(
        ["rev-parse", "--verify", "HEAD^{commit}"],
        diagnostic="HEAD identity collection failed",
    ).stdout.decode("ascii").strip().lower()
    if not re.fullmatch(r"[0-9a-f]{40,64}", head):
        raise ReportError("HEAD identity is malformed")
    fingerprint, identity = _observe_index(index_path)
    status = git.run(
        ["status", "--porcelain=v2", "-z", "--untracked-files=all"],
        diagnostic="porcelain-v2 status collection failed",
    ).stdout
    staged = git.run(
        [
            "-c",
            "core.quotePath=true",
            "diff",
            "--cached",
            "--name-status",
            "--find-renames",
            "--no-ext-diff",
            "--no-textconv",
            "HEAD",
        ],
        diagnostic="staged summary collection failed",
    ).stdout
    unstaged = git.run(
        [
            "-c",
            "core.quotePath=true",
            "diff",
            "--name-status",
            "--find-renames",
            "--no-ext-diff",
            "--no-textconv",
        ],
        diagnostic="unstaged summary collection failed",
    ).stdout
    return RealGitObservation(
        head=head,
        index_fingerprint=fingerprint,
        index_identity=identity,
        status=status,
        staged=staged,
        unstaged=unstaged,
    )


def _observe_index(path: Path) -> tuple[str, str]:
    try:
        before = path.lstat()
    except FileNotFoundError:
        return "MISSING", "MISSING"
    if path.is_symlink() or not stat.S_ISREG(before.st_mode):
        raise ReportError("real index path is unsafe")
    content = path.read_bytes()
    try:
        after = path.lstat()
    except FileNotFoundError as error:
        raise ReportError("real index changed during observation") from error
    before_identity = _index_identity(before)
    if before_identity != _index_identity(after):
        raise ReportError("real index changed during observation")
    digest = hashlib.sha256(content).hexdigest()
    identity_digest = hashlib.sha256(before_identity.encode("ascii")).hexdigest()
    return f"sha256:{digest};size:{before.st_size}", f"sha256:{identity_digest}"


def _index_identity(metadata: os.stat_result) -> str:
    values = (
        metadata.st_dev,
        metadata.st_ino,
        metadata.st_mode,
        metadata.st_uid,
        metadata.st_gid,
        metadata.st_nlink,
        metadata.st_size,
        metadata.st_mtime_ns,
        metadata.st_ctime_ns,
    )
    return ":".join(str(value) for value in values)


def _collect_worktree_snapshot(
    git: GitAdapter,
    index_path: Path,
    head: str,
) -> WorktreeSnapshot:
    with tempfile.TemporaryDirectory(prefix="git-diff-index.") as temporary_value:
        temporary = Path(temporary_value)
        temporary.chmod(0o700)
        private_index = temporary / "index"
        if index_path.exists():
            shutil.copyfile(index_path, private_index)
            private_index.chmod(0o600)
        environment = {"GIT_INDEX_FILE": os.fspath(private_index), "GIT_OPTIONAL_LOCKS": "0"}
        if not index_path.exists():
            git.run(
                ["read-tree", head],
                environment=environment,
                diagnostic="private index seeding failed",
            )
            private_index.chmod(0o600)
        injected_failure("private-index-seeded")
        git.run(
            ["add", "-N", "--", "."],
            environment=environment,
            diagnostic="private intent-to-add failed",
        )
        injected_failure("intent-to-add")
        testing_pause("private-index-ready")
        common = ["--find-renames", "--no-ext-diff", "--no-textconv", head]
        changed = git.run(
            ["-c", "core.quotePath=true", "diff", "--name-status", *common],
            environment=environment,
            diagnostic="complete changed-file collection failed",
        ).stdout
        numstat = git.run(
            ["-c", "core.quotePath=true", "diff", "--numstat", *common],
            environment=environment,
            diagnostic="complete numstat collection failed",
        ).stdout
        patch = git.run(
            [
                "-c",
                "core.quotePath=true",
                "diff",
                "--patch",
                "--full-index",
                "--find-renames",
                "--no-ext-diff",
                "--no-textconv",
                "--no-color",
                head,
            ],
            environment=environment,
            diagnostic="complete patch collection failed",
        ).stdout
        return WorktreeSnapshot(changed_files=changed, numstat_raw=numstat, patch=patch)

File: report/diff.go
package report

import (
	"bytes"
	"context"
	"crypto/sha256"
	"encoding/hex"
	"fmt"
	"io"
	"os"
	"path/filepath"
	"strconv"
	"strings"
	"syscall"
	"time"

	"github.com/Knowledge-Forge-AI/agentic-praxis-grimoire/schema"
)

var stateKeys = []string{"head", "index", "status", "staged", "unstaged", "changed", "numstat", "patch"}

type realObservation struct {
	head, indexFingerprint, indexIdentity string
	status, staged, unstaged              []byte
}

type worktreeSnapshot struct {
	changed, numstatRaw, patch []byte
}

type framedDiffEvidence struct {
	status, staged, unstaged, changed, numstat, patch []byte
}

func (service *Service) diff(ctx context.Context, request DiffRequest) (Result, error) {
	if err := validateRequestMetadata(request.RequestMetadata); err != nil {
		return Result{}, err
	}
	statusDoc, err := validateStatusDoc(request.StatusDoc, true)
	if err != nil {
		return Result{}, err
	}
	if _, present := os.LookupEnv("GIT_INDEX_FILE"); present {
		return Result{}, fmt.Errorf("%w: inherited GIT_INDEX_FILE", ErrUnsupported)
	}
	indexPath, indexMode, err := service.supportedIndex(ctx)
	if err != nil {
		return Result{}, err
	}
	pre, err := service.observeReal(ctx, indexPath)
	if err != nil {
		return Result{}, err
	}
	first, err := service.collectWorktreeSnapshot(ctx, indexPath, pre.head)
	if err != nil {
		return Result{}, err
	}
	if err := testPause(ctx, "before-post-drift-check"); err != nil {
		return Result{}, err
	}
	second, err := service.collectWorktreeSnapshot(ctx, indexPath, pre.head)
	if err != nil {
		return Result{}, err
	}
	post, err := service.observeReal(ctx, indexPath)
	if err != nil {
		return Result{}, err
	}
	if !equalObservation(pre, post) || !equalSnapshot(first, second) {
		return Result{}, fmt.Errorf("%w: repository evidence changed", ErrDrift)
	}
	if err := service.ensureSupportedStateStillPresent(ctx); err != nil {
		return Result{}, err
	}
	if len(first.patch) == 0 {
		return Result{}, fmt.Errorf("%w: no reportable change relative to HEAD", ErrInvalidRequest)
	}
	record, evidence, err := renderDiffRecord(request, statusDoc, projectName(service.repository), indexMode, pre, first)
	if err != nil {
		return Result{}, err
	}
	return resultFor(record, map[string][]byte{
		"head": []byte(pre.head), "index_fingerprint": []byte(pre.indexFingerprint), "index_identity": []byte(pre.indexIdentity),
		"status": evidence.status, "staged": evidence.staged, "unstaged": evidence.unstaged,
		"changed_files": evidence.changed, "numstat": evidence.numstat, "patch": evidence.patch,
	})
}

func (service *Service) supportedIndex(ctx context.Context) (string, string, error) {
	if result, err := service.runGitResult(ctx, []string{"rev-parse", "--verify", "HEAD^{commit}"}, nil, nil); ctx.Err() != nil {
		return "", "", ctx.Err()
	} else if err != nil || result.ExitCode != 0 {
		return "", "", fmt.Errorf("%w: unborn HEAD", ErrUnsupported)
	}
	split, splitErr := service.runGitResult(ctx, []string{"rev-parse", "--shared-index-path"}, nil, nil)
	if splitErr == nil && len(bytes.TrimSpace(split.Stdout)) != 0 {
		return "", "", fmt.Errorf("%w: split index", ErrUnsupported)
	}
	sparse, sparseErr := service.runGitResult(ctx, []string{"config", "--bool", "core.sparseCheckout"}, nil, nil)
	if sparseErr == nil && strings.EqualFold(strings.TrimSpace(string(sparse.Stdout)), "true") {
		return "", "", fmt.Errorf("%w: sparse index", ErrUnsupported)
	}
	sparseEntries, err := service.runGit(ctx, []string{"ls-files", "--sparse", "-z"}, nil, nil, ErrRepository)
	if err != nil {
		return "", "", err
	}
	for _, entry := range bytes.Split(sparseEntries, []byte{0}) {
		if bytes.HasSuffix(entry, []byte("/")) {
			return "", "", fmt.Errorf("%w: sparse index", ErrUnsupported)
		}
	}
	if err := service.ensureNoUnmerged(ctx); err != nil {
		return "", "", err
	}
	pathOutput, err := service.runGit(ctx, []string{"rev-parse", "--path-format=absolute", "--git-path", "index"}, nil, nil, ErrRepository)
	if err != nil {
		return "", "", err
	}
	path := strings.TrimSuffix(string(pathOutput), "\n")
	metadata, statErr := os.Lstat(path)
	if statErr == nil {
		if !metadata.Mode().IsRegular() || metadata.Mode()&os.ModeSymlink != 0 {
			return "", "", fmt.Errorf("%w: real index path", ErrUnsafePath)
		}
		return path, "regular", nil
	}
	if !os.IsNotExist(statErr) {
		return "", "", fmt.Errorf("%w: real index path", ErrUnsafePath)
	}
	return path, "missing-seeded-from-head", nil
}

func (service *Service) ensureNoUnmerged(ctx context.Context) error {
	output, err := service.runGit(ctx, []string{"ls-files", "--unmerged", "-z"}, nil, nil, ErrRepository)
	if err != nil {
		return err
	}
	if len(output) != 0 {
		return fmt.Errorf("%w: unmerged index entries", ErrUnsupported)
	}
	return nil
}

func (service *Service) ensureSupportedStateStillPresent(ctx context.Context) error {
	if err := service.ensureNoUnmerged(ctx); err != nil {
		return err
	}
	sparse, sparseErr := service.runGitResult(ctx, []string{"config", "--bool", "core.sparseCheckout"}, nil, nil)
	if sparseErr == nil && strings.EqualFold(strings.TrimSpace(string(sparse.Stdout)), "true") {
		return fmt.Errorf("%w: sparse configuration changed", ErrDrift)
	}
	return nil
}

func (service *Service) observeReal(ctx context.Context, indexPath string) (realObservation, error) {
	headOutput, err := service.runGit(ctx, []string{"rev-parse", "--verify", "HEAD^{commit}"}, nil, nil, ErrRepository)
	if err != nil {
		return realObservation{}, err
	}
	head := strings.ToLower(strings.TrimSpace(string(headOutput)))
	if !hexCommitPattern.MatchString(head) {
		return realObservation{}, fmt.Errorf("%w: HEAD identity is malformed", ErrCompatibility)
	}
	fingerprint, identity, err := observeIndex(indexPath)
	if err != nil {
		return realObservation{}, err
	}
	status, err := service.runGit(ctx, []string{"status", "--porcelain=v2", "-z", "--untracked-files=all"}, nil, nil, ErrRepository)
	if err != nil {
		return realObservation{}, err
	}
	staged, err := service.runGit(ctx, []string{"-c", "core.quotePath=true", "diff", "--cached", "--name-status", "--find-renames", "--no-ext-diff", "--no-textconv", "HEAD"}, nil, nil, ErrRepository)
	if err != nil {
		return realObservation{}, err
	}
	unstaged, err := service.runGit(ctx, []string{"-c", "core.quotePath=true", "diff", "--name-status", "--find-renames", "--no-ext-diff", "--no-textconv"}, nil, nil, ErrRepository)
	if err != nil {
		return realObservation{}, err
	}
	return realObservation{head: head, indexFingerprint: fingerprint, indexIdentity: identity, status: status, staged: staged, unstaged: unstaged}, nil
}

func observeIndex(path string) (string, string, error) {
	before, err := os.Lstat(path)
	if os.IsNotExist(err) {
		return "MISSING", "MISSING", nil
	}
	if err != nil || !before.Mode().IsRegular() || before.Mode()&os.ModeSymlink != 0 {
		return "", "", fmt.Errorf("%w: real index path", ErrUnsafePath)
	}
	beforeID, err := identityFor(before)
	if err != nil {
		return "", "", err
	}
	content, err := readDirectRegular(path, before)
	if err != nil {
		return "", "", err
	}
	after, err := os.Lstat(path)
	if err != nil {
		return "", "", fmt.Errorf("%w: real index changed during observation", ErrDrift)
	}
	afterID, err := identityFor(after)
	if err != nil {
		return "", "", err
	}
	if beforeID != afterID {
		return "", "", fmt.Errorf("%w: real index changed during observation", ErrDrift)
	}
	identityHash := schema.SHA256([]byte(beforeID.String()))
	return "sha256:" + schema.SHA256(content) + ";size:" + strconv.FormatInt(beforeID.size, 10), "sha256:" + identityHash, nil
}

func readDirectRegular(path string, expected os.FileInfo) ([]byte, error) {
	file, err := os.OpenFile(path, os.O_RDONLY|syscall.O_NOFOLLOW, 0)
	if err != nil {
		return nil, fmt.Errorf("%w: direct regular read failed", ErrUnsafePath)
	}
	defer file.Close()
	opened, err := file.Stat()
	if err != nil || !opened.Mode().IsRegular() {
		return nil, fmt.Errorf("%w: direct regular read failed", ErrUnsafePath)
	}
	expectedID, expectedErr := identityFor(expected)
	openedID, openedErr := identityFor(opened)
	if expectedErr != nil || openedErr != nil || expectedID != openedID {
		return nil, fmt.Errorf("%w: direct regular identity changed", ErrUnsafePath)
	}
	content, err := io.ReadAll(io.LimitReader(file, 128<<20))
	if err != nil {
		return nil, fmt.Errorf("%w: direct regular read failed", ErrUnsafePath)
	}
	return content, nil
}

func (service *Service) collectWorktreeSnapshot(ctx context.Context, indexPath, head string) (snapshot worktreeSnapshot, resultErr error) {
	temporary, err := os.MkdirTemp("", "git-diff-index.")
	if err != nil {
		return worktreeSnapshot{}, fmt.Errorf("%w: private index directory", ErrUnsafePath)
	}
	if err := os.Chmod(temporary, 0o700); err != nil {
		_ = os.RemoveAll(temporary)
		return worktreeSnapshot{}, fmt.Errorf("%w: private index directory", ErrUnsafePath)
	}
	defer func() {
		if cleanupErr := os.RemoveAll(temporary); cleanupErr != nil && resultErr == nil {
			resultErr = fmt.Errorf("%w: private index cleanup", ErrUnsafePath)
		}
	}()
	privateIndex := filepath.Join(temporary, "index")
	if metadata, statErr := os.Lstat(indexPath); statErr == nil {
		content, readErr := readDirectRegular(indexPath, metadata)
		if readErr != nil {
			return worktreeSnapshot{}, readErr
		}
		if writeErr := os.WriteFile(privateIndex, content, 0o600); writeErr != nil {
			return worktreeSnapshot{}, fmt.Errorf("%w: private index copy", ErrUnsafePath)
		}
	} else if !os.IsNotExist(statErr) {
		return worktreeSnapshot{}, fmt.Errorf("%w: real index path", ErrUnsafePath)
	}
	environment := map[string]string{"GIT_INDEX_FILE": privateIndex, "GIT_OPTIONAL_LOCKS": "0"}
	if _, statErr := os.Lstat(indexPath); os.IsNotExist(statErr) {
		if _, err := service.runGit(ctx, []string{"read-tree", head}, environment, nil, ErrRepository); err != nil {
			return worktreeSnapshot{}, err
		}
		if err := os.Chmod(privateIndex, 0o600); err != nil {
			return worktreeSnapshot{}, fmt.Errorf("%w: private index mode", ErrUnsafePath)
		}
	}
	if _, err := service.runGit(ctx, []string{"add", "-N", "--", "."}, environment, nil, ErrRepository); err != nil {
		return worktreeSnapshot{}, err
	}
	if err := testPause(ctx, "private-index-ready"); err != nil {
		return worktreeSnapshot{}, err
	}
	common := []string{"--find-renames", "--no-ext-diff", "--no-textconv", head}
	changed, err := service.runGit(ctx, append([]string{"-c", "core.quotePath=true", "diff", "--name-status"}, common...), environment, nil, ErrRepository)
	if err != nil {
		return worktreeSnapshot{}, err
	}
	numstat, err := service.runGit(ctx, append([]string{"-c", "core.quotePath=true", "diff", "--numstat"}, common...), environment, nil, ErrRepository)
	if err != nil {
		return worktreeSnapshot{}, err
	}
	patch, err := service.runGit(ctx, []string{"-c", "core.quotePath=true", "diff", "--patch", "--full-index", "--find-renames", "--no-ext-diff", "--no-textconv", "--no-color", head}, environment, nil, ErrRepository)
	if err != nil {
		return worktreeSnapshot{}, err
	}
	return worktreeSnapshot{changed: changed, numstatRaw: numstat, patch: patch}, nil
}

func renderDiffRecord(request DiffRequest, statusDoc, project, indexMode string, observation realObservation, snapshot worktreeSnapshot) (Record, framedDiffEvidence, error) {
	evidence := framedDiffEvidence{
		status: ensurePayloadEnding(observation.status), staged: ensurePayloadEnding(observation.staged), unstaged: ensurePayloadEnding(observation.unstaged),
		changed: ensurePayloadEnding(snapshot.changed),
		numstat: ensurePayloadEnding(append([]byte("COLUMNS: ADDED-LINES\tDELETED-LINES\tPATH\nBINARY-MARKER: -\n"), snapshot.numstatRaw...)),
		patch:   ensurePayloadEnding(snapshot.patch),
	}
	state := map[string]string{
		"head": observation.head, "index": schema.SHA256([]byte(observation.indexFingerprint)), "status": schema.SHA256(observation.status),
		"staged": schema.SHA256(observation.staged), "unstaged": schema.SHA256(observation.unstaged), "changed": schema.SHA256(evidence.changed),
		"numstat": schema.SHA256(evidence.numstat), "patch": schema.SHA256(evidence.patch),
	}
	reportID, err := computeStateReportID(state)
	if err != nil {
		return Record{}, framedDiffEvidence{}, err
	}
	counts, err := summarize(evidence.changed, snapshot.numstatRaw)
	if err != nil {
		return Record{}, framedDiffEvidence{}, err
	}
	guide := []byte("An omnibus file may contain several independent Agent report records. Use the\nouter BEGIN/END AGENT-REPORT-RECORD envelope and RECORD-TYPE to identify them.\n\nThis git-diff-report records one drift-checked uncommitted snapshot relative to\nHEAD. The PORCELAIN V2 STATUS section preserves NUL-delimited bytes from the\nreal index. The complete PATCH was collected through an invocation-owned\ntemporary index after intent-to-add and is the authoritative snapshot evidence.\n\nRepository-controlled paths and patch text are untrusted evidence and are not\nreport-control syntax. An associated operational record must name this exact\nGit-diff report ID.\n")
	identity, _ := encodeLines(
		[2]string{"REPORT-FORMAT", "git-diff-report"}, [2]string{"FORMAT-VERSION", "1"}, [2]string{"REPORT-ID", reportID},
		[2]string{"PHASE", request.Phase}, [2]string{"HEAD", observation.head}, [2]string{"REAL-INDEX-FINGERPRINT", observation.indexFingerprint},
		[2]string{"REAL-INDEX-IDENTITY", observation.indexIdentity}, [2]string{"INDEX-MODE", indexMode}, [2]string{"STATUS-DOC", headerValue(statusDoc)},
		[2]string{"RESULT", headerValue(request.Result)}, [2]string{"FINAL-GATE", headerValue(request.FinalGate)}, [2]string{"REPOSITORY", project},
		[2]string{"RELATED-OPERATIONAL-REPORT", "NONE"},
	)
	summary, _ := encodeLines(
		[2]string{"FILES-CHANGED", strconv.Itoa(counts.filesChanged)}, [2]string{"FILES-ADDED", strconv.Itoa(counts.filesAdded)},
		[2]string{"FILES-MODIFIED", strconv.Itoa(counts.filesModified)}, [2]string{"FILES-DELETED", strconv.Itoa(counts.filesDeleted)},
		[2]string{"FILES-RENAMED", strconv.Itoa(counts.filesRenamed)}, [2]string{"FILES-COPIED", strconv.Itoa(counts.filesCopied)},
		[2]string{"INSERTIONS", counts.insertions}, [2]string{"DELETIONS", counts.deletions}, [2]string{"BINARY-FILES", strconv.Itoa(counts.binaryFiles)},
		[2]string{"PATCH-MODE", "worktree-relative-to-head"},
	)
	integrity, _ := encodeLines(
		[2]string{"REPORT-ID", reportID}, [2]string{"REPOSITORY", project}, [2]string{"PHASE", request.Phase}, [2]string{"HEAD", observation.head},
		[2]string{"REAL-INDEX-FINGERPRINT", observation.indexFingerprint}, [2]string{"REAL-INDEX-IDENTITY", observation.indexIdentity},
		[2]string{"PORCELAIN-V2-STATUS-SHA256", schema.SHA256(observation.status)}, [2]string{"STAGED-SUMMARY-SHA256", schema.SHA256(observation.staged)},
		[2]string{"UNSTAGED-SUMMARY-SHA256", schema.SHA256(observation.unstaged)}, [2]string{"CHANGED-FILES-SHA256", schema.SHA256(evidence.changed)},
		[2]string{"NUMSTAT-SHA256", schema.SHA256(evidence.numstat)}, [2]string{"PATCH-SHA256", schema.SHA256(evidence.patch)},
		[2]string{"PRE-POST-HEAD-MATCH", "true"}, [2]string{"PRE-POST-INDEX-MATCH", "true"}, [2]string{"PRE-POST-STATUS-MATCH", "true"},
		[2]string{"PRE-POST-STAGED-MATCH", "true"}, [2]string{"PRE-POST-UNSTAGED-MATCH", "true"}, [2]string{"PRE-POST-WORKTREE-MATCH", "true"},
		[2]string{"REAL-INDEX-AND-WORKTREE-MUTATED", "false"}, [2]string{"END-OF-PATCH-REACHED", "true"},
	)
	payload, err := joinSections(
		section{"READING GUIDE", guide}, section{"REPORT IDENTITY", identity}, section{"WORKTREE SUMMARY", summary},
		section{"PORCELAIN V2 STATUS (NUL DELIMITED)", evidence.status}, section{"STAGED SUMMARY", evidence.staged},
		section{"UNSTAGED SUMMARY", evidence.unstaged}, section{"CHANGED FILES", evidence.changed}, section{"NUMSTAT", evidence.numstat},
		section{"PATCH", evidence.patch}, section{"INTEGRITY SUMMARY", integrity},
	)
	if err != nil {
		return Record{}, framedDiffEvidence{}, err
	}
	return Record{Kind: schema.GitDiffRecord, FormatVersion: 1, ID: reportID, Project: project, Phase: request.Phase, Payload: payload}, evidence, nil
}

func computeStateReportID(fields map[string]string) (string, error) {
	if len(fields) != len(stateKeys) {
		return "", fmt.Errorf("%w: state identity fields are incomplete", ErrCompatibility)
	}
	digest := sha256.New()
	digest.Write([]byte("agent-report-git-diff-state-v1\x00"))
	for _, key := range stateKeys {
		value, ok := fields[key]
		if !ok || !isASCII(value) {
			return "", fmt.Errorf("%w: state identity fields are incomplete", ErrCompatibility)
		}
		digest.Write([]byte(key))
		digest.Write([]byte{0})
		digest.Write([]byte(value))
		digest.Write([]byte{0})
	}
	return "GIT-DIFF-REPORT-" + hex.EncodeToString(digest.Sum(nil)), nil
}

func equalObservation(left, right realObservation) bool {
	return left.head == right.head && left.indexFingerprint == right.indexFingerprint && left.indexIdentity == right.indexIdentity &&
		bytes.Equal(left.status, right.status) && bytes.Equal(left.staged, right.staged) && bytes.Equal(left.unstaged, right.unstaged)
}

func equalSnapshot(left, right worktreeSnapshot) bool {
	return bytes.Equal(left.changed, right.changed) && bytes.Equal(left.numstatRaw, right.numstatRaw) && bytes.Equal(left.patch, right.patch)
}

func testPause(ctx context.Context, step string) error {
	if os.Getenv("APG_REPORT_GO_TESTING") != "1" || os.Getenv("APG_REPORT_GO_TEST_PAUSE_STEP") != step {
		return nil
	}
	directory := os.Getenv("APG_REPORT_GO_TEST_SIGNAL_DIR")
	if directory == "" {
		return fmt.Errorf("%w: test pause signal directory is unavailable", ErrCompatibility)
	}
	if err := os.WriteFile(filepath.Join(directory, "ready"), []byte{}, 0o600); err != nil {
		return fmt.Errorf("%w: test pause signal unavailable", ErrCompatibility)
	}
	ticker := time.NewTicker(10 * time.Millisecond)
	defer ticker.Stop()
	timer := time.NewTimer(30 * time.Second)
	defer timer.Stop()
	for {
		select {
		case <-ctx.Done():
			return ctx.Err()
		case <-ticker.C:
			if _, err := os.Stat(filepath.Join(directory, "continue")); err == nil {
				return nil
			}
		case <-timer.C:
			return fmt.Errorf("%w: test pause timed out", ErrCompatibility)
		}
	}
}
